feat: add cheatsheet

theborakompanioni opened the pull request:
This PR adds a cheatsheet component.

It is currently only shown in `devMode` for now, as some items on the list (e.g. Fidelity Bonds), are not yet supported.

Differences to the Figma design:
- Adds a close button to the overlay - this is needed on small mobile devices (Should it be hidden on bigger screens?)
- Footer Link Icon does not match (currently using `File` as placeholder – found no real alternative on https://bitcoinicons.com/ yet)

## 📸
<img src="https://user-images.githubusercontent.com/3358649/162630840-4177792e-a89e-4f3f-80c7-8584689c64a4.png" width="400"><img src="https://user-images.githubusercontent.com/3358649/162630834-3cd40559-f48b-43b0-9189-38c01e6452df.png" width="400">

theborakompanioni commented:
Footer `Nav.Item`'s still need some styling on mobile.

editwentyone commented:
Figma: added a close button and changed icon to the normal bitcoin file icon. valid point for mobile use.
 
would love to see the cheatsheet jump up after a delay when the user uses JAM for the first time, directly after the creation of the first wallet (see rapid prototype for a better understanding). 

also its better to hide "cheatsheet" in the footer until the user created the first wallet (if easy and possible). just to not confuse too much at the beginning.

Djo1e commented on src/components/Cheatsheet.tsx:
```diff
@@ -0,0 +1,136 @@
+  return <div className="numbered">{number}</div>
+}
+
+function ListItem({ number, children }: PropsWithChildren<NumberedProps>) {
```
I think you can take this one step further

```
interface ListItemProps {
  number: number
  transaltionKey: string
  descriptionKey: string
}

function ListItem({ number, children, transaltionKey, descriptionKey }: PropsWithChildren<ListItemProps>) {
  const { t } = useTranslation()
  return (
    <rb.Stack className="cheatsheet-list-item" direction="horizontal" gap={3}>
      <div className="numbered">{number}</div>
      <rb.Stack gap={0}>
        <h6>
          <Trans i18nKey={transaltionKey}>{children}</Trans>
        </h6>
        <div className="small text-secondary">{t(descriptionKey)}</div>
      </rb.Stack>
    </rb.Stack>
  )
}
```

```
<ListItem number={1} transaltionKey="cheatsheet.item_1.title" descriptionKey="cheatsheet.item_1.description">
    <Link to="/receive">Fund</Link> your wallet first.
</ListItem>

```

theborakompanioni replied:
Passing translation keys strikes me as unnecessarily restrictive.
It is so little code - feels like over-engineering to me for now. Do not get me wrong - I _love_ to over-engineer.. : D
But feel free to add any improvements once the PR is merged!

theborakompanioni commented on src/index.css:
```diff
@@ -357,6 +357,51 @@ main {
   }
 }
 
+/* Cheatsheet Styles */
```
Yep, that's a good idea. I actually tried and could not make it work for `.tsx` files. I think there is something missing and it needs a little more thought on how to do it properly - so I refrained from it in this PR. If you have a decent solution already in your mind, please keep it coming!

Djo1e replied:
I ran into the same issue actually. Adding https://github.com/joinmarket-webui/joinmarket-webui/blob/cad7ef9203ad71fe8c58e42f919609fbc2242dce/typings.d.ts this should fix it

theborakompanioni replied:
Nice - I think we will do a refactoring session once this gets integrated - there a few more components where this comes in handy.. :orange_heart:

theborakompanioni commented:
Updated screenshots.

theborakompanioni commented:
> would love to see the cheatsheet jump up once fully automatically after a delay when the user uses JAM for the first time, [...]

Done. Will be shown once after a wallet has been unlocked, just like the onboarding screen.

> 
> the underlined words are links, the first 3 links in the text are linking to openoms sources and the other links are linking to parts inside the jam app. i.e.: (1) **Fund** your wallet first. (click it inside the prototype)

:+1:

> also its better to hide "cheatsheet" in the footer until the user created the first wallet (if easy and possible). just to not confuse too much at the beginning.

It will be displayed only if a wallet is unlocked - not on the login screen (links would not work).
> 
> I gave point 4 fidelity bond another color to show its inactive for now, but never the less, there are some informations that enlightens the user. maybe its a good idea to also communicate the upcoming feature FB

:+1: Done. Opacity decreased for item 4.

dergigi commented:
Nice! That looks great already!

Some word choices could be improved, or at least we should explain the concepts somewhere. Here is what came to mind:

* Why is "adding noise" good?
* What is "sweep"?
* "very private" - how about "more private"?
* Not sure if we should focus on yield that heavily, as others mentioned in the past the amounts that you will earn are probably rather low. Maybe do some expectation management?

Feel free to ignore the above, we can improve all this at a later point as well. It's a good first version! 👏

theborakompanioni commented:
> Nice! That looks great already!

All honor to @editwentyone.

> Some word choices could be improved, or at least we should explain the concepts somewhere. 

I am always grateful for concrete wording suggestions, because there are so many people who have the necessary talent for this - but I am afraid I am not one of them.

> Here is what came to mind:
> 
>     * Why is "adding noise" good?

Changed from "add more noise" to "increase privacy of all". "noise" is avoided entirely. It is not used anywhere and would - like you said - need an explanation. Acceptable approach?

>     * What is "sweep"?

Since we use the term "sweep" on the Send screen also, do you think it can be explained there and is this the appropriate context? Refrained from changing anything in the cheatsheet component regarding this for now.

>     * "very private" - how about "more private"?

:+1: "more private" it is.

>     * Not sure if we should focus on yield that heavily, as others mentioned in the past the amounts that you will earn are probably rather low. Maybe do some expectation management?

Yep, that is true. Maybe also better explained on the "Earn" screen, when the context is appropriate (?).
> 
> Feel free to ignore the above, we can improve all this at a later point as well. It's a good first version! clap

One question before it is merged - should item 3 and 4 be flipped?

dergigi commented:
> Should item 3 and 4 be flipped?

I'd say yes.

Another thing that touches on something that we discussed often—something that isn't fully resolved yet—is the mentioning of "levels." 

If we manage to implement and ship #179 with the next release, manual consolidation (and thus anything that mentions "levels" at all) might fall to the wayside for most users, since this would be automated with scheduled transactions.

I'll propose some wording changes in the code directly.

dergigi commented on src/components/Cheatsheet.tsx:
```diff
@@ -0,0 +1,143 @@
+          <rb.Offcanvas.Title>{t('cheatsheet.title')}</rb.Offcanvas.Title>
+          <div className="small text-secondary">
+            <Trans i18nKey="cheatsheet.description">
+              This is one valid way to add random noise to your coins by switching from{' '}
```
```suggestion
              Follow the steps below to increase your financial privacy. It is advisable to switch from{' '}
```

dergigi commented on src/components/Cheatsheet.tsx:
```diff
@@ -0,0 +1,143 @@
+          <ListItem number={1}>
+            <h6>
+              <Trans i18nKey="cheatsheet.item_1.title">
+                <Link to={routes.receive}>Fund</Link> your wallet first.
```
```suggestion
                <Link to={routes.receive}>Fund</Link> your wallet.
```

dergigi commented:
I just noticed that there are two sources of truth for the text due to the translation engine, so I will just type it out here before I propose code changes that might be misplaced:

---

**The Cheatsheet**

Follow the steps below to increase your financial privacy. It is advisable to switch from [earning as a maker](#) to [sending as a taker](#) back and forth. [Learn more.](#)

**1) Fund your wallet.**
Deposit some funds yourself or receive it from others.

**2) Send a collaborative transaction to yourself.**
Collaborative transactions increase everyone's privacy.

**3) Optional: Lock funds in a fidelity bond.**
A fidelity bond increases your chances of selling liquidity substantially.

**4) Provide liquidity and earn yield.**
Offer your sats to the marketplace. No trust or custody required—you are always in full control of your funds.

**5) Make use of scheduled transactions**
Automatically plan and execute a series of collaborative transactions. You can also automatically sweep your funds to cold storage, or use them to open a lightning channel, for example.

**6) Go to step one and repeat.**
Still confused? Dig into the documentation.

---

This assumes that we will ship #179 soon, which I think we will. I'm an optimist! 😁 

Let me know what you think.

theborakompanioni commented:
> I just noticed that there are two sources of truth for the text due to the translation engine, so I will just type it out here before I propose code changes that might be misplaced.

Ah, yes, totally forgot that fact when applying the suggestions before.

> This assumes that we will ship #179 soon, which I think we will. I'm an optimist! grin

Yes. Worst case is, we have to apply css class `upcoming-feature` to that item. Not an issue :+1:

> Let me know what you think.

I'd merge it! Ping @editwentyone for a competent design perspective.

![image](https://user-images.githubusercontent.com/3358649/162977405-ba4196c8-9ffd-40b5-8c04-0aff52731546.png)
![image](https://user-images.githubusercontent.com/3358649/162977445-cb487886-cdc2-4955-84fb-7d9d1224df7e.png)

editwentyone commented:
hey :) because on mobile its a lot of text, I would suggest to fix the Numbers at the top of the cell, I also fixed it in Figma. 

also I would suggest to link only the important words like Fund, Send, Earn, it reads faster

theborakompanioni commented:
> hey :) because on mobile its a lot of text, I would suggest to fix the Numbers at the top of the cell, I also fixed it in Figma.

:+1: Done.

> also I would suggest to link only the important words like Fund, Send, Earn, it reads faster

:+1: Done.
I changed to wording slightly in order for links come first (at least in English).
- "Provide liquidity and earn yield." => "Earn yield by providing liquidity."
- "Make use of scheduled transactions." => "Use scheduled transactions."

<img src="https://user-images.githubusercontent.com/3358649/163137898-0230f638-ef3c-41b8-9593-010752df1d6d.png" width=350 />
<img src="https://user-images.githubusercontent.com/3358649/163137930-53b2a205-e921-4d65-910f-b68c11a11a4d.png" width=400 />

dergigi reviewed (approved):
Looks good to me ✅

editwentyone commented:
one little annoyance: by clicking a link like "Fund", its switching the page in the background and keeps the cheatsheet still in the front. could we close the cheatsheet also with the same click? 

(+ is it possible to overlay the complete page including the header when the cheatsheet is active?)

theborakompanioni commented:
> one little annoyance: by clicking a link like "Fund", its switching the page in the background and keeps the cheatsheet still in the front. could we close the cheatsheet also with the same click?

I thought of that as a feature instead of a bug : D
But of course, it is possible! Would you open a ticket for that? Could be tagged with "good first issue".

> (+ is it possible to overlay the complete page including the header when the cheatsheet is active?)

This is a little bit trickier, but also possible of course. The reason is that the navbar should be shown as usual on some occasions when a backdrop is active, e.g. when showing the navigation on mobile. Personally, if I may say that, I'd rather not do that.
